app/models/csv_dumps/dump_processor.rb

```diff
@@ -0,0 +1,62 @@
+module CsvDumps
+  class DumpProcessor
+    include ActiveSupport::Callbacks
+
+    attr_reader :csv_dump, :formatter, :scope, :medium
+
+    define_callbacks :dump
+    define_callbacks :upload
+    define_callbacks :cleanup
+
+    def initialize(formatter, scope, medium, csv_dump = CsvDump.new)
+      @formatter = formatter
+      @scope = scope
+      @medium = medium
+      @csv_dump = csv_dump
+    end
+
+    def execute
+      perform_dump
+      upload_dump
+    ensure
+      cleanup_dump
+    end
+
+    def perform_dump
+      csv_dump << formatter.headers if formatter.headers
+
+      read_from_database do
+        scope.each do |model|
+          formatter.to_rows(model).each do |row|
+            csv_dump << row
+          end
+        end
+      end
+    end
+
+    def upload_dump
+      gzip_file_path = csv_dump.gzip!
+      write_to_s3(gzip_file_path)
+      set_ready_state
+    end
+
+    def cleanup_dump
+      csv_dump.cleanup!
+    end
+
+    private
+
+    def set_ready_state
+      medium.metadata["state"] = 'ready'
+      medium.save!
+    end
+
+    def write_to_s3(gzip_file_path)
+      medium.put_file(gzip_file_path, compressed: true)
+    end
+
+    def read_from_database(&block)
```

Comment:
Seems like this is duplicating the underlying scope each methods (e.g. `CsvDumps::DumpScope`) that specify to `read_from_database` too. Maybe one or the other to ensure we read from the replica